packages/material-ui/src/utils/useId.js

```diff
@@ -0,0 +1,20 @@
+import * as React from 'react';
+
+/**
+ * Private module reserved for @material-ui/x packages.
+ */
+export default function useId(idProp) {
+  const [defaultId, setDefaultId] = React.useState();
+  const id = idProp || defaultId;
+  React.useEffect(() => {
+    if (defaultId) {
```

Comment:
This should be strict. The behavior is otherwise a bit weird on 0, '' and especially switching between null and undefined.

Reply:
This was lifted from existing code :) But actually, I think that unless I derive state from `idProp` it would never get a non string value due to the stringification on `Math.random()` below. Would you like a `!= null` check?

Reply:
Also I think currently components type `id` and `name` as `PropTypes.string` so you shouldn't get numbers.

Reply:
Yeah I know we do this in a lot of places. I'll try to prevent introducing more. I consider these passing as a bug in eslint. If we disallow loose equality we should also disallow type coearcion in `if (mixedType)`. 

It's one of those things where code is optimized for writing and not reading and I've seen too many bugs due to type coercion.

Reply:
So would you rather `!= null` here or what? And should I make the change to initialise state with `idProp`?